pkg/vault/test/foundry/utils/BaseExtremeAmountsTest.sol

```diff
@@ -0,0 +1,546 @@
+                userData: bytes("")
+            })
+        );
+        assertEq(bptAmountIn, bptAmountOut, "bptAmountIn should be equal to bptAmountOut");
+
+        // if amountsOut[0] more then amountsIn[0], then amountsOut[1] should be less than amountsIn[1] otherwise user gets free tokens
+        // if amountsOut[0] equal to amountsIn[0], then amountsOut[1] should be less or equal to amountsIn[1] otherwise user gets free tokens
+        // if amountsOut[0] less then amountsIn[0], we can't check anything because user doesn't get free tokens
+        if (amountsOut[0] > amountsIn[0]) {
+            assertLt(amountsOut[1], amountsIn[1], "amountsOut[1] should be less than amountsIn[1]");
+        } else if (amountsOut[0] == amountsIn[0]) {
+            assertLe(amountsOut[1], amountsIn[1], "amountsOut[1] should be less or equal to amountsIn[1]");
+        }
+    }
+
+    // testAddProportionalAndRemoveLiquidityExactIn
+    function testAddProportionalAndRemoveLiquidityExactIn_Fuzz(
+        uint256 exactBPTAmount,
+        uint256[2] memory balancesRaw,
+        uint256 swapFee
+    ) public {
+        _setUpVariables(_boundBalances(balancesRaw), _boundSwapFee(swapFee));
+
+        _testAddProportionalAndRemoveLiquidityExactIn(_boundExactBPTAmountForSingleTokenOperations(exactBPTAmount));
+    }
+
+    function testAddProportionalAndRemoveLiquidityExactIn_FuzzSwapFee(uint256 swapFee) public {
+        swapFee = _boundSwapFee(swapFee);
+        _setUpVariables([HUGE_INIT_AMOUNT, HUGE_INIT_AMOUNT].toMemoryArray(), swapFee);
+
+        _testAddProportionalAndRemoveLiquidityExactIn(MIN_AMOUNT);
+    }
+
+    function testAddProportionalAndRemoveLiquidityExactIn_FuzzBPTAmount(uint256 exactBPTAmount) public {
+        _setUpVariables([HUGE_INIT_AMOUNT, HUGE_INIT_AMOUNT].toMemoryArray(), maxSwapFee / 2);
+
+        _testAddProportionalAndRemoveLiquidityExactIn(_boundExactBPTAmountForSingleTokenOperations(exactBPTAmount));
+    }
+
+    function _boundExactBPTAmountForSingleTokenOperations(uint256 exactBPTAmount) private view returns (uint256) {
+        return bound(exactBPTAmount, MIN_AMOUNT, _calculateMaxBPTAmountForSingleTokenOperations());
+    }
+
+    function _testAddProportionalAndRemoveLiquidityExactIn(uint256 exactBPTAmount) private {
+        vault.forceUnlock();
+
+        (uint256[] memory amountsIn, uint256 bptAmountOut, ) = vault.addLiquidity(
+            AddLiquidityParams({
+                pool: pool,
+                to: address(this),
+                maxAmountsIn: [maxAdditionalAmountsIn[0], maxAdditionalAmountsIn[1]].toMemoryArray(),
+                minBptAmountOut: exactBPTAmount,
+                kind: AddLiquidityKind.PROPORTIONAL,
+                userData: bytes("")
+            })
+        );
+        assertEq(bptAmountOut, exactBPTAmount, "bptAmountOut should be equal to exactBPTAmount");
+
+        uint removePerOperation = exactBPTAmount / 2;
+
+        // first single token exact in removal
```

Comment:
let's just clarify that we can't remove everything with a single token because there's not enough balance to do so (is that right?)

Reply:
the idea of ​​this test is to remove tokenA first and then tokenB and check what the difference is with proportional removal. so we take the BPT amount from the proportional addition and divide by two